### ISSUE TITLE
1.9.3 tests fail on FreeBSD

### ISSUE THREAD

**Issue by mexas** [sourceryinstitute/OpenCoarrays#475]
Please help

http://eis.bris.ac.uk/~mexas/log/opencoarrays193-freebsd.log

1.9.1 passed all tests.
I haven't tried 1.9.2.

**Comment by vehre** [comment]
This happens to me, when the mpi-library can not be found on the LD_LIBRARY_PATH. Build/testscript deficiency? @zbeekman

**Comment by mexas** [comment]
Wrong MPI library is used:
```
-- Found MPI_C: /usr/local/mpi/openmpi/lib/libmpi.so (found version "3.0") 
-- Found MPI_Fortran: /usr/local/mpi/openmpi/lib/libmpi_usempi.so (found version "3.0") 
```
It should be using MPICH.

I need to find where this is coming from.
The port gives the user a choice of mpich, openmpi and openmpi2.
mpich is the default, which is what I'm testing now.

**Comment by mexas** [comment]
I think
```
find_package( MPI )
```
is to blame.
I don't know which MPI it chooses when multiple MPI libraries
are installed on the system. For some reason it chooses openmpi.
And of course, it doesn't know of the user's chosen MPI option,
when building opencoarrays.

**Comment by mexas** [comment]
yes, disabling find_package(MPI) helps:
```
-- Found MPI_C: /usr/local/lib/libmpi.so (found version "3.1") 
-- Found MPI_Fortran: /usr/local/lib/libmpifort.so (found version "3.1") 
```
100% tests pass:

<http://eis.bris.ac.uk/~mexas/log/opencoarrays193-freebsd-mpich.log>

**Comment by mexas** [comment]
However, now I cannot build with openmpi...
I need to find out how FreeBSD port options
are passed to opencoarrays cmake build

**Comment by vehre** [comment]
Well, usually it is the other way around. You can supply to find_package(MPI) which mpi-installation to choose. See for example here: https://cmake.org/cmake/help/v3.10/module/FindMPI.html
But I don't know how to pass this to the coarray installer. May be --help can give a clue.

**Comment by zbeekman** [comment]
Building packages for distribution should be done by invoking cmake directly, which I suspect is what @mexas is doing. I haven’t had a chance to look at the logs and the issue template was not followed. Recent versions of CMake have substantial changes to FindMPI. (One of the reasons why we ask for the CMake version in the issue template)

I’ll try to look at the attached logs; hopefully they contain all the required info to debug this issue.

**Comment by mexas** [comment]
```
# cmake --version
cmake version 3.10.1
```

**Comment by zbeekman** [comment]
@mexas your log does not show how CMake is being invoked during the build. I need this information to determine why it is picking up the wrong MPI.

**Comment by zbeekman** [comment]
> Btw, you are using clang for the c-compiler and gfortran for the fortran parts. I think this is unsupported.

@vehre I would categorize clang use as unofficially supported at this point. By default on macOS when using Homebrew clang is used as well, and I have not seen any ill effects. Perhaps we should be officially supporting it. As soon as I can add some CI tests with clang we can change it to be officially supported.

**Comment by mexas** [comment]
@zbeekman if I invoke make with "-dx", I get this:
```
===>  Performing out-of-source build
/bin/mkdir -p /usr/ports/lang/opencoarrays/work/.build
+ cd /usr/ports/lang/opencoarrays/work/.build
+ /usr/bin/env 'F77=gfortran6' 'FC=gfortran6' 'FFLAGS=-O -Wl,-rpath=/usr/local/lib/gcc6' 'FCFLAGS=-W
l,-rpath=/usr/local/lib/gcc6' 'XDG_DATA_HOME=/usr/ports/lang/opencoarrays/work' 'XDG_CONFIG_HOME=/us
r/ports/lang/opencoarrays/work' 'HOME=/usr/ports/lang/opencoarrays/work' 'PATH=/usr/ports/lang/openc
oarrays/work/.bin:/sbin:/bin:/usr/sbin:/usr/bin:/usr/local/sbin:/usr/local/bin:/root/bin' 'SHELL=/bi
n/sh' 'CONFIG_SHELL=/bin/sh' /usr/local/bin/cmake '-DCMAKE_C_COMPILER:STRING=cc' '-DCMAKE_CXX_COMPIL
ER:STRING=c++' '-DCMAKE_C_FLAGS:STRING=-O2 -pipe  -fstack-protector -fno-strict-aliasing' '-DCMAKE_C
_FLAGS_DEBUG:STRING=-O2 -pipe  -fstack-protector -fno-strict-aliasing' '-DCMAKE_C_FLAGS_RELEASE:STRI
NG=-O2 -pipe  -fstack-protector -fno-strict-aliasing' '-DCMAKE_CXX_FLAGS:STRING=-O2 -pipe -fstack-pr
otector -fno-strict-aliasing ' '-DCMAKE_CXX_FLAGS_DEBUG:STRING=-O2 -pipe -fstack-protector -fno-stri
ct-aliasing ' '-DCMAKE_CXX_FLAGS_RELEASE:STRING=-O2 -pipe -fstack-protector -fno-strict-aliasing ' '
-DCMAKE_EXE_LINKER_FLAGS:STRING= -Wl,-rpath=/usr/local/lib/gcc6  -L/usr/local/lib/gcc6 -B/usr/local/
bin -fstack-protector' '-DCMAKE_MODULE_LINKER_FLAGS:STRING= -Wl,-rpath=/usr/local/lib/gcc6  -L/usr/l
ocal/lib/gcc6 -B/usr/local/bin -fstack-protector' '-DCMAKE_SHARED_LINKER_FLAGS:STRING= -Wl,-rpath=/u
sr/local/lib/gcc6  -L/usr/local/lib/gcc6 -B/usr/local/bin -fstack-protector' '-DCMAKE_INSTALL_PREFIX
:PATH=/usr/local' '-DCMAKE_BUILD_TYPE:STRING=Release' '-DTHREADS_HAVE_PTHREAD_ARG:BOOL=YES' '-DCMAKE
_INSTALL_RPATH_USE_LINK_PATH=YES' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' /usr/ports/lang/opencoarrays/wo
rk/OpenCoarrays-1.9.3
-- The C compiler identification is Clang 4.0.0
```
Is this the info you were looking for?

The full log:
<http://eis.bris.ac.uk/~mexas/log/1.log>

**Comment by zbeekman** [comment]
@mexas Yes this looks like it has all the info I was looking for.

A few follow on questions:

1. Is the only patching that's occurring truncating the CMAKE_Fortran_COMPILER_VERSION to the major version for the `mod_dir_tail`?
(This is a change that *might* be worthwhile to consider upstreaming into our build system. My only concern is that module files between minor versions may have incompatibilities, maybe @vehre knows if GFortran makes any guarantees about this.)

2. When you say "disabling find_package(MPI) helps" presumably you are deleting it in the top level CMakeLists.txt only? It is called again in src/mpi/CMakeLists.txt but the call in the top level CMakeLists.txt is mostly a work around/accomodation for the easy installation script with the deficiencies in the older FindMPI.cmake.

If the answer to Q2 is "yes" then this helps greatly, because we now know that the call in src/mpi/CMakeLists.txt is being handled correctly but the top level one is not.

**Comment by mexas** [comment]
@zbeekman yes to both

Q1: Well, disabling find_packages(MPI) is the second patch. In the build log I show you it wasn't applied. 

RE: compiler_version: Using the full gcc version triplet breaks the FreeBSD pkg system, because they use only the major version, i.e. there is an env var defined for this.

Q2: yes, only the top level CMakeLists.txt. However, now that you ask, I see that for 1.9.1 it was different. For 1.9.1 the patch was:
```
@${REINPLACE_CMD} '/mod_install/s/$${CMAKE_Fortran_COMPILER_VERSION}/${_GCC_VER}/' \
               ${WRKSRC}/src/mpi/CMakeLists.txt
```

For 1.9.3 I went for:

```
@${REINPLACE_CMD} '/mod_dir_tail/s/$${CMAKE_Fortran_COMPILER_VERSION}/${_GCC_VER}/' \
               ${WRKSRC}/CMakeLists.txt
```

Maybe that's why it's broken between 1.9.1 and 1.9.3...

**Comment by mexas** [comment]
sorry, typed too fast. I now see that what I wrote is confusing.

disabling find_packages(MPI) was never in 1.9.1. This is a new patch for 1.9.3.

It's the patch that truncated the GCC version to the major version, that is different between 1.9.1 and 1.9.3. In 1.9.1 I applied it to src/mpi/CMakeLists.txt and in 1.9.3 I apply it to the top level CMakeLists.txt.

**Comment by zbeekman** [comment]
> Maybe that's why it's broken between 1.9.1 and 1.9.3...

Your initial instinct that it has to do with find_package(MPI) smells right to me. Still investigating. I'll apply a fix to truncate to the major version in the next release (at least on BSD systems, if not others too)

**Comment by zbeekman** [comment]
@mexas We're hoping to get a release out soon. I'd like to address this issue if at all possible.

You said:
> I don't know which MPI it chooses when multiple MPI libraries
> are installed on the system. For some reason it chooses openmpi.
> And of course, it doesn't know of the user's chosen MPI option,
> when building opencoarrays.

How were you specifying one MPI vs another in the past?
Also can you please show me the patch where you removed `find_package( MPI )` in 1.9.3? `find_package(MPI)` or some variant is called in a few different places depending on some other installation logic.

> However, now I cannot build with openmpi...

Can you elaborate more?

FYI, recent versions of mpiexec shipped with OpenMPI require a hosts file, which was causing tests to fail w/ OpenMPI in the past, but now we detect and provide a hosts file for running the tests with ctest. (But only for the local machine.)

The canonical way to tell CMake where MPI is appears to be by passing `-DMPI_<lang>_COMPILER` to point to the MPI C and Fortran wrappers when calling CMake. However, just setting up your path so that one or the other is available should also work.

**Comment by mexas** [comment]
@zbeekman:
"How were you specifying one MPI vs another in the past?"

Well, via FreeBSD port options framework:

```
OPTIONS_SINGLE=		MPI
OPTIONS_SINGLE_MPI=	MPICH OPENMPI OPENMPI2
OPTIONS_DEFAULT=	MPICH
```

However, I never made it to pass tests with OpenMPI - this is a problem with FreeBSD OpenMPI port.
I have a PR submitted to the OpenMPI FreeBSD port maintainer to look into this.
So - maybe it never worked...

About find_package(MPI) patch:

<https://bz-attachments.freebsd.org/attachment.cgi?id=189655>

specifically:
```
's/find_package( MPI )/#find_package( MPI )/'  ${WRKSRC}/CMakeLists.txt
```
where ${WRKSRC} expands into OpenCoarrays-1.9.3, so it's the top level
CMakeLists.txt file.

"However, now I cannot build with openmpi..." - basically with my patch the version of MPI found is MPICH. Without the patch the version of MPI found is OpenMPI. This is wrong, as it should be managed solely by FreeBSD port options: default MPICH, but if the user selected OpenMPI2, then use it. Perhaps FreeBSD MPI options conflict with find_package(MPI)? Perhaps find_package(MPI) overwrites the FreeBSD ports options. I admit I don't know exactly how FreeBSD port options propagate - I'll need somebody with more ports experience to get into that.

I need to check with somebody in FreeBSD if -DMPI_<lang>_COMPILER is used or not.

**Comment by mexas** [comment]
I got some info from the FreeBSD ports team. I'm checking a few things.

One question - what should "caf -w" return - gfortran* or mpif90?

**Comment by mexas** [comment]
Testing with openmpi2 I got this build error:
```
[ 12%] Generating hello_multiverse
cd /usr/ports/lang/opencoarrays/work/.build/src/tests/integration/coarrayHelloWorld && ../../../../bin/caf -I /usr/local/mpi/openmpi2/include -I /usr/local/mpi/openmpi2/lib -DPREFIX_NAME=_gfortran_caf_ -DGCC_GE_7 -DMPI_WORKING_MODULE -O3 -DNDEBUG -O3 -o /usr/ports/lang/opencoarrays/work/.build/bin/OpenCoarrays-1.9.3-tests/hello_multiverse /usr/ports/lang/opencoarrays/work/OpenCoarrays-1.9.3/src/tests/integration/coarrayHelloWorld/hello_multiverse.F90
../../../../bin/caf: line 87: syntax error near unexpected token `;'
*** [src/tests/integration/coarrayHelloWorld/hello_multiverse] Error code 2
```
The full log:
<http://eis.bris.ac.uk/~mexas/log/opencoarrays193_openmpi2.log>

**Comment by mexas** [comment]
with openmpi, I get opencoarrays built/installed, only 1 test passes:
```
      Start 47: test-installation-scripts.sh
47/47 Test #47: test-installation-scripts.sh .....   Passed    0.20 sec

2% tests passed, 46 tests failed out of 47

```
but I can use it run simple coarray code:
```
20474701626e> caf -s
/usr/local/bin/gfortran7 -I/usr/local/include/OpenCoarrays-1.9.3_GNU-7 -fcoarray=lib -pthread -Wl,-rpath=-Wl,-rpath=/usr/local/lib/gcc7 -Wl,-rpath -Wl,/usr/local/mpi/openmpi/lib -Wl,--enable-new-dtags -pthread ${@} /usr/local/lib/libcaf_mpi.a /usr/local/mpi/openmpi/lib/libmpi_usempi.so /usr/local/mpi/openmpi/lib/libmpi_mpifh.so /usr/local/mpi/openmpi/lib/libmpi.so
20474701626e> cat z.f90 
implicit none
write (*,*) this_image(), num_images()
end
20474701626e> caf z.f90 
20474701626e> cafrun -np 7 ./a.out
           2           7
           4           7
           6           7
           7           7
           1           7
           5           7
           3           7
20474701626e> 
```

**Comment by mexas** [comment]
Forgot to say, I reverted back the find_package(MPI) patch.
The only patch I'm using now is the GCC version:
```
        @${REINPLACE_CMD} '/mod_dir_tail/s/$${CMAKE_Fortran_COMPILER_VERSION}/${_GCC_VER}/' \
                ${WRKSRC}/CMakeLists.txt
```

**Comment by mexas** [comment]
Finally, with MPICH, I can build/install 1.9.3, and run a simple program:
```
20474701626e> caf -s
/usr/local/bin/gfortran7 -I/usr/local/include/OpenCoarrays-1.9.3_GNU-7 -fcoarray=lib -Wl,-rpath=/usr/local/lib/gcc7 -Wl,-rpath -Wl,/usr/local/lib ${@} /usr/local/lib/libcaf_mpi.a /usr/local/lib/libmpifort.so /usr/local/lib/libmpi.so
20474701626e> cat z.f90 
implicit none
write (*,*) this_image(), num_images()
end
20474701626e> caf z.f90 
20474701626e> cafrun -np 7 ./a.out
           1           7
           4           7
           5           7
           7           7
           2           7
           3           7
           6           7
20474701626e> 
```
However, every time I run "make test" (only when built with MPICH), all tests pass and then...
my laptop reboots.

I'll report this to FreeBSD, in case the problem is on their end.

**Comment by zbeekman** [comment]
@mexas let me try to catch up a little bit

---

> Well, via FreeBSD port options framework: ...

I need to know what affect those settings in the port options framework have on the build environment... What environment variables are set, how the PATH variable changes, are packages sym-linked/un-sym-linked etc. I have no experience with FreeBSD ports, so I need to know what the practical effects of those settings are.

---

> One question - what should "caf -w" return - gfortran* or mpif90?

This should return `gfortran*`. The CMake `FindMPI` module extracts all the compilation & link flags, definitions, libraries, etc. needed. Then we build the `caf` and `cafrun` scripts so that they have the libraries and link flags that were detected at OpenCoarrays configure, and the compiler detected during configuration. This is because the OC RTL has certain features and interfaces that are dependent on both the version of GFortran as well as the capabilities of the MPI implementation. On my MacOS system the two useful flags for debugging `caf` are:

```bash
$ bin/caf --wraps
caf wraps /usr/local/bin/gfortran-7
# and
$ bin/caf --show
/usr/local/bin/gfortran-7 -I/Users/ibeekman/Sandbox/OpenCoarrays/build/include/OpenCoarrays-1.9.3-159-g0ade739_GNU-7.3.0 -fcoarray=lib -Wl,-flat_namespace -Wl,-commons,use_dylibs ${@} /Users/ibeekman/Sandbox/OpenCoarrays/build/lib/libcaf_mpi.a /usr/local/Cellar/mpich/3.2.1_1/lib/libmpifort.dylib /usr/local/Cellar/mpich/3.2.1_1/lib/libmpi.dylib /usr/local/Cellar/mpich/3.2.1_1/lib/libpmpi.dylib
```

Can you please run `bin/caf --show` and `bin/cafrun --show` from the build directory before running `make test` or `ctest`? Also, if you can install the [shellcheck](https://github.com/koalaman/shellcheck) package that should trigger some additional sanity tests on the `caf` and `cafrun` scripts.

---

> Testing with openmpi2 I got this build error: `../../../../bin/caf: line 87: syntax error near unexpected token ';'`

This looks like the caf script is either not getting properly configured (due to CMake stuff), or the arguments being passed to it in the test are messed up (due CMake stuff). If you could attach or paste lines 48-95 of the generated bin/caf script that would be very helpful. Between `# Begin configured variables` to `# End configured variables`. It looks like CMake is injecting a semi-colon separated list rather than a space separated list. I will guard against this in the CMakeLists.txt file

---

> with openmpi, I get opencoarrays built/installed, only 1 test passes:

Can you please do one or more of the following when running the tests

* Run `make check`
* Run `ctest --output-on-failure`
* Set the environment variable `CTEST_OUTPUT_ON_FAILURE` to the value `TRUE` (only supported by more recent versions of CMake)

The intrinsic target `make test` created by cmake does not show output on failure. This will give much more information.

One issue I have fixed on master is that OpenMPI 3.x was requiring a hosts file to run the tests now. So it's possible that some of your OpenMPI woes will be solved on master.

---

> Forgot to say, I reverted back the find_package(MPI) patch.

Does this mean that at least MPICH is being found correctly? Or are all versions of MPI now being found correctly?

Regarding finding the correct MPI: I noticed in [this log](http://eis.bris.ac.uk/~mexas/log/1.log) (by the way please always include the `-dx` flags, it helps substantially) the text: 

```
+ /usr/bin/env 'dp_RAWDEPENDS=bash:shells/bash mpif90:net/mpich ...
```

If this is one way of specifying the MPICH vs OpenMPI dependency, then a better executable to specify here (or in addition to mpif90) would be `mpirun` or `mpiexec`. The location of the MPI launcher is the first thing examined when attempting to locate the MPI version to be used with the FindMPI module.

---

> However, every time I run "make test" (only when built with MPICH), all tests pass and then...
> my laptop reboots.
> 
> I'll report this to FreeBSD, in case the problem is on their end.

Yikes! Which version of MPICH? Also, you might try reconfiguring and building with MPICH & `-DCAF_ENABLE_FAILED_IMAGES=FALSE` and see if that helps the spontaneous reboot issue.

**Comment by zbeekman** [comment]
@mexas, feel free to respond to each of my individual responses above; however, at this point, this thread is pretty hard to follow and understand. We have a new 2.0.0-rc1 release (https://github.com/sourceryinstitute/OpenCoarrays/releases/tag/2.0.0-rc1) that should fix one or two of the myriad of issues reported throughout this thread.

I recommend that you try the new 2.0.0-rc1 release candidate and that we close this issue. Then, please, please use the issue template, and limit the scope of each issue to each individual problem that you encounter when testing 2.0.0-rc1.

I definitely admire and appreciate your enthusiasm, and we are all pleased to have OpenCoarrays available as a FreeBSD port, but, without a FreeBSD machine to debug with, we should be a little bit more surgical when addressing the issues you are seeing. This way there is hope of resolving them.

Thanks

**Comment by mexas** [comment]
ok
